File: src/clj/curatify/tasks/analyze.clj
(ns curatify.tasks.analyze
  (:require [curatify.db.core :as db]
            [clojure.tools.logging :as log]))

(defn playlist-affinities [playlist-id]
  (let  [artist-affinities (db/playlist-artist-affinity {:playlist-id playlist-id})
         genre-affinities (db/playlist-genre-affinity {:playlist-id playlist-id})
         total (:count (db/playlist-track-count {:playlist-id playlist-id}))]
    (as-> artist-affinities aa
         (reduce
           (fn [acc {:keys [artist_id count name]}]
             (assoc acc artist_id {:name name
                                   :percentage (/ count total)}))
           {}
           aa)
         (db/update-artist-affinities! {:playlist-id playlist-id :aa aa}))
    (as-> genre-affinities ga
          (reduce
            (fn [acc {:keys [genre count]}]
              (assoc acc genre (/ count total)))
            {}
            ga)
          (db/update-genre-affinities! {:playlist-id playlist-id :ga ga}))))


(defn track-affinities [playlist]
  (let [playlist-id (:id playlist)
        {genre-aff :genre_affinities artist-aff :artist_affinities} (db/get-playlist-affinities {:playlist-id playlist-id})
        user-id (:user_id playlist)
        tracks (db/get-user-unheard-inbox {:id user-id})]
    (doall (map (fn [track]
                  (let [genre-overlap (or (select-keys genre-aff (map keyword (:genres track))) {})
                        track-genre-aff (or (reduce + (vals genre-overlap)) 0)
                        artist-overlap (or (select-keys artist-aff (map keyword (:artists track))) {})
                        track-artist-aff (or (reduce + (map :percentage (vals artist-overlap))) 0)
                        total-aff (+ track-artist-aff track-genre-aff)]
                    (db/set-inbox-track-affinities {:user-id user-id
                                                    :track-id (:id track)
                                                    :playlist-aff (merge (:playlist_affinities track)
                                                                        {playlist-id total-aff})})))
                tracks))))



(defn analyze-all []
  (log/info "Analyzing all playlist affinities")
  (let [playlists (db/get-curated-playlists)]
    (log/info (with-out-str (time (doall (map playlist-affinities (map :id playlists))))))
    (log/info (with-out-str (time (doall (map track-affinities playlists)))))))

File: server/tasks/analyze.js
import {
  updatePlaylistArtistAffinities,
  updatePlaylistGenreAffinities,
  allPlaylists,
  userCuratedPlaylists
} from '../data/playlists'
import { userUnheardInbox, updateTrackPlaylistMatches } from '../data/inbox'
import { allUsers } from '../data/users'

export async function analyzeAll () {
  const playlists = await allPlaylists()

  await Promise.all(playlists.map(async p => {
    await updatePlaylistArtistAffinities(p)
    await updatePlaylistGenreAffinities(p)
  }))

  const users = await allUsers()

  await Promise.all(users.map(async user => {
    console.info(`analyzing for user ${user.id}...`)
    await calculateTrackPlaylistMatches(user)
  }))
}

async function calculateTrackPlaylistMatches (user) {
  const playlists = await userCuratedPlaylists(user.id)
  const unheard = await userUnheardInbox(user)

  await Promise.all(unheard.map(async track => {
    const match = Object.max(playlists.reduce((acc, playlist) => {
      const genreScore = Object.values(
        Object.select(playlist.genre_affinities, track.genres)
      ).sum()

      const artistScore = Object.values(
        Object.select(playlist.artist_affinities, track.artists)
      ).sum()
      return {
        [playlist.id]: artistScore + genreScore,
        ...acc
      }
    }, {}))
    await updateTrackPlaylistMatches(match, track.track_id, user.id)
  }))
}
